trie/fog.py

```diff
@@ -0,0 +1,284 @@
+import ast
+from itertools import zip_longest
+from typing import (
+    Any,
+    Sequence,
+    Tuple,
+)
+
+from eth_utils import (
+    to_tuple,
+    ValidationError,
+)
+from sortedcontainers import SortedSet
+
+from trie.typing import (
+    Nibbles,
+    NibblesInput,
+    RawHexaryNode,
+)
+from trie.utils.nodes import (
+    key_starts_with,
+)
+
+
+class HexaryTrieFog:
+    """
+    Keeps track of which parts of a trie have been verified to exist.
+
+    Named after "fog of war" popular in video games like... Red Alert? IDK, I'm old.
+
+    Object is immutable. Any changes, like marking a key prefix as complete, will
+    return a new HexaryTrieFog object.
+    """
+    _unexplored_prefixes: SortedSet
+
+    # INVARIANT: No unexplored prefix may start with another unexplored prefix
+    #   For example, _unexplored_prefixes may not be {(1, 2), (1, 2, 3)}.
+
+    def __init__(self) -> None:
+        # Always start without knowing anything about a trie. The only unexplored
+        #   prefix is the root prefix: (), which means the whole trie is unexplored.
+        self._unexplored_prefixes = SortedSet({()})
+
+    def __repr__(self) -> str:
+        return f"HexaryTrieFog<{self._unexplored_prefixes!r}>"
+
+    @property
+    def is_complete(self) -> bool:
+        return len(self._unexplored_prefixes) == 0
+
+    def explore(
+            self,
+            old_prefix: NibblesInput,
+            sub_segments_input: Sequence[NibblesInput]) -> 'HexaryTrieFog':
+        """
+        The fog lifts from the old prefix. This call returns a HexaryTrieFog that narrows
+        down the unexplored key prefixes. from the old prefix to the indicated children.
+
+        For example, if only the key prefix 0x12 is unexplored, then calling
+        explore((1, 2), ((3,), (0xe, 0xf))) would mark large swaths of 0x12 explored, leaving only
+        two prefixes as unknown: 0x123 and 0x12ef. To continue exploring those prefixes, navigate
+        to them using traverse() or traverse_from().
+
+        The sub_segments_input may be empty, which means the old prefix has been fully explored.
+        """
+        sub_segments = [Nibbles(segment) for segment in sub_segments_input]
+        new_fog_prefixes = self._unexplored_prefixes.copy()
+
+        try:
+            new_fog_prefixes.remove(old_prefix)
+        except KeyError:
+            raise ValidationError(f"Old parent {old_prefix} not found in {new_fog_prefixes!r}")
+
+        if len(set(sub_segments)) != len(sub_segments):
+            raise ValidationError(
+                f"Got duplicate sub_segments in {sub_segments} to HexaryTrieFog.explore()"
+            )
+
+        # Add further validation that no segment is a prefix of another
+        all_lengths = set(len(segment) for segment in sub_segments)
+        if len(all_lengths) > 1:
+            for segment in sub_segments:
+                shorter_lengths = [length for length in all_lengths if length < len(segment)]
```

Comment:
It seems like we should be able to do *better* if we modeled the data as a tree.  IIUC the current approach is super-linear in runtime complexity.

```python
tree = {}
for segment in sub_segments:
    sub_tree = tree
    for nibble in segment:
        sub_tree.setdefault(nibble, {})
        if not isinstance(sub_tree[nibble], collections.Mapping):
            ...  # we've detected a that `segment` is a superset of `sub_tree[nibble]`
        sub_tree = sub_tree[nibble]
    # once the segment terminates, we store the full segment at that position 
    # so that we can later detect segments that are prefixes of other segments.
    sub_tree[segment[-1]] = segment
```

I think the untested psuedocode above has `O(n)` complexity.

Reply:
I'm not too worried about speed here because the whole thing is behind the `if len(all_lengths) > 1:` check, which won't actually trigger for any use cases we care about (branch, extension, or leaf nodes being explored one at a time).

I'll add a comment to that ^ effect, and open an issue to come back to later.